Feature(app): Reactify FCC!

BerkeleyTrue opened the pull request:
- [x] build javascript/html view
- [x] build step view
- [x] add data fetching
- [x] add test suite rendering from data
- [x] build out challengeFramework
  - [x] convert to CJS
  - [x] run tests in iframe
  - [x] run preview/tests/output in separate frames
  - [x] console logging to output view
- [x] add map view
- [x] ~~video~~ [now called project] challenge view
- [x] Submit/next challenge toast
- [x] move to next challenge
### BETA ready
- [x] move Videos under same folder as other challenges (in progress)
- [x] move back a challenge
- [x] add language routing
- [x] move hints from challenge articles to JSON files
  - [x] define JSON hint structure
  - [x] create react hint elements and redux actions for enabling
- [x] fix map view (drawer broken)
- [x] add LightBox to step challenges
- [x] add map view search/filter
- [x] add hot keys
- [ ] generate code URI on request
- [x] add GA url tracking
### QA (Bring down the hamma!)
- [ ] Fix all da bugs!
### Post merge
- [ ] Modern challenge interface

~~MDN links~~ MDN links will be moved to the hint system
### Changes that will need to made to all challenges
- Challenges that use to implicitly display function output will no longer do so. This challenges will need to use a console log to display that information to the user.

closes #4464
closes #5072
closes #5111
closes #6607
closes #6608
closes #6609
closes #6910 
closes #7831 
closes #8290 
closes #9157

Commit:
Add react challenge view with editor

Commit:
Initial step challenge ui with state changes

Commit:
Move to localize non ui data

Commit:
Add challenge map service

Commit:
Normalize model json using generator loopback

Commit:
Add webpack cold reloading
On changes to the react bundle
webpack will store the current redux state
in localStorage, waits (to allow the server to restart)
then refreshes the page. On page load, it checks if it
has state stored and loads it into the app.

Commit:
Only use file hashing in production

Commit:
Add webpack code splitting module
Add cold-module replacement
Add webpack module hashing

Commit:
Add React Map

Commit:
Break up map components
Add static super blocks

Commit:
remove active class on map button

Commit:
Add logic to filter input field
Add correct css to input field addon icon

Commit:
Add filter clearing logic

Commit:
Redirect /challenges to /map

Commit:
Make map links react aware

Commit:
Add new rx saga

Commit:
Move to redux-epic

jsJunky commented on client/sagas/err-saga.js:
```diff
@@ -1,20 +1,14 @@
+  return action$
+    .filter(({ error }) => !!error)
+    .map(({ error }) => error)
+    .doOnNext(error => console.error(error))
```
:question: Can you also just use `.doOnNext(console.error)` or is that not valid?

BerkeleyTrue replied:
You could, but you would need to bind the method to the console object
```js
.doOnNext(console.error.bind(console))
```

jsJunky replied:
Yeah that looks pretty gross. Why does referencing the console object behave differently when outside of the doOnNext scope? I'll play around with that this weekend Thanks for responding!

BerkeleyTrue replied:
In the doOnNext method call you are passing in a function. If you pass in the method from the console object it loses it's internal context (the `this` object internally becomes the `this` object of the callee function). 

Binding it to the console object ensures that its internal context cannot change.

jsJunky replied:
Ah so it auto binds this much like arrow functions. Thanks!

Commit:
Initial work on new framework

Commit:
Move fetchChallenges up a level

Commit:
Add initial dynamic challenge rendering

Commit:
Create polyvinyl type

Commit:
Move selectors around

Commit:
move map as a supdirectory of challenges

Commit:
update current challenge on map click

Commit:
Make PolyVinyl's functional

Commit:
Move content into files

Commit:
Fix window height adjustment during dev

Commit:
fix file storing for single file challenges

Commit:
Fix content decoding

Commit:
Add bonfire to challenge types

Commit:
Add dynamic tests

Commit:
Update file on codemirror change

Commit:
Move Old style challenge to classic

Commit:
Add debounce to editor

Commit:
Add head/tail to file

Commit:
Update rx ajax typings

Commit:
Initial preview rendering

Commit:
Rename folder

Commit:
Run tests in iframe displays results to user

Commit:
Challenge now display console.log in output window

Commit:
Remove unused files

Commit:
Add code storage on code execution

Commit:
Update tests to use pass/err to determine refresh

Commit:
Add random compliment on challenge completion

Commit:
Actually require invariant

Commit:
Fix production webpack build

Commit:
Fix lint errors
Remove commonFramework build from gulpfile

Commit:
Moves to next challenges

Commit:
Fix challenge ordering

Commit:
Add map drawer
This commit break map css. Will need to refactor how the map renders

Commit:
Add extensible views to challenges

Commit:
Add initial project view

Commit:
Add main chat toggle

Commit:
Fix test check typo
Fix punctuation

Commit:
Move event method calls out of action creator

Commit:
Change project list according to #8867

Commit:
Add jsx closing bracket rule

Commit:
Add view logic for all projects

Commit:
Submit projects

Commit:
Submit simple project (previously known as video challegnes)

Commit:
Submit step

Commit:
Submit step challenge

Commit:
Remove step pointer events block
unminify animate.css

Commit:
Add block scoping to challenges url

Commit:
Dashify challenge block property on seed

Commit:
Move from one challenge to another within a block

Commit:
Fix next challenge loading logic

Grimmreaper81 commented on client/frame-runner.js:
```diff
@@ -0,0 +1,73 @@
+  var chai = parent.chai;
+  var source = document.__source;
+
+  document.__getJsOutput = function getJsOutput() {
```
Spread(

Commit:
Move Video challenges under challenges dir
Remove old hikes components
Remove unused jobs stuff

Commit:
Add tern-project file

Commit:
Move back to previous challenge loads new challenge

Commit:
Update redux-actions

Commit:
Udpate react-router-redux

Commit:
Update(package): React-youtube

Commit:
Feature(challenges): Add hint system to challenges

Commit:
feature(translations): Normalize translations
translations are now listed under the prop "translations"
and keys to the IETF lang tag.

Commit:
Add language routing

Commit:
Feature(vim): add libs to tern
Refactor map services;

Commit:
Feature(react): Add lightbox to step challenge images

Commit:
Feature(mousetrap): Add mousetrap features to redux

Commit:
Feature(components): fetch user after page load
This makes it easier to serve whole site statically in the future
Feature(redux): Move user state into entities

Commit:
Feature(redux): Add completed challenge logic to redux

Commit:
fix(meta): Remove tern-project from tracking

Commit:
Feature(langauge): Make client history language aware
Remove need for language aware links

Commit:
Fix(map): prevent loading a challenge from refetching

Commit:
Fix(challenges): Reset challenge ui on unmount

Commit:
Refactor(challenges): refactor challenge completion saga

Commit:
Fix(challenges): reset project form on submit

Commit:
Fix(challenges): load next block on block completion

Commit:
Refactor(map): Fix lint issues

Commit:
Fix(map): update map filter components and esc key event

Commit:
Refactor(map): move map header to own component

Commit:
Feature(map): Add top level collapse

Commit:
feature(map): Add collapse to block level

Commit:
feature(map): move locked static challenges to seed files

Commit:
Fix(map): disable map transition animation
Fix(map): block/challenge vertical alignment

Commit:
Feature(map): add expand/collapse all button/logic

Commit:
Feature(map): change caret when open

Commit:
Fix(map): Fix superblock title/name on api call

Commit:
fix(drawer): enable scroll on map drawer

Commit:
Fix(map): structure to work with new drawer

Commit:
Feature(map): add super block message logic

Commit:
Fix(lint): correct lint errors

Commit:
Fix(lint): Add import eslint and fix import errors

Commit:
Fix(css): move iphone image down a scooch
closes #9339

Commit:
Fix(map): remove map drawer from non-react pages
map drawer will be reserved for react pages only
closes #9341

Commit:
Refactor(views): remove old challenge/map jade files

Commit:
Fix(challenge): Add spacer underneath testing suite
closes #9342

Commit:
Fix(server): remove legacy redirect
closes #9345

Commit:
Fix(map): prevent map drawer from opening on map route

Commit:
Fix(map): Fix map drawer responsiveness
close #9348

Commit:
Fix(challenges): make challenge height auto
closes #9347

Commit:
Fix(challenge): coming soon challenges unclickable in prod

Commit:
Add(cover): add cover(alls) scripts and javascript testing

Commit:
Add(tests): test out next challenge utils

Commit:
Add(challenges): Add dev mode exception to next challenge logic

Commit:
Feature(challenges): Move to coming soon challenges in dev
close #9349

Commit:
Fix(views): remove wiki/map iframe drawers
closes #9460

Commit:
Fix: typo in super block name
closes #9451

Commit:
Fix(challenges): scroll instruction panel to top between challenges
closes #9361

Commit:
fix(server): typo in commit router
closes #9345

Commit:
Feature(toast): Move from react-toastr to react-notifications

Commit:
Feature(toast): fix toast positioning

Commit:
Feature(challenges): add reset logic

Commit:
Feature(toasts): make toast short lived by default

Commit:
Feature(toast): remove on action click

Commit:
Feature(chat): Add help chat logic

Commit:
Feature(challenges): add bug modal and logic

Commit:
Fix(package): Bad copy paste in test script

Commit:
Feature(map): Filter challenges on Map
closes #9346

Commit:
Feature(pages): Add 404 page
To prevent loops, server errors will lead here instead of map page
This page can be static to prevent infinite loops
closes #9463

Commit:
Feature(react): Move settings to React

Commit:
Feature(nav): Image/Points leads to settings
This keeps the user in React land when changing settings
User profiles will be reactified at a later date

Commit:
Feature(settings): add user flag logic
This also moves a couple of settings to their own controller

Commit:
Feature(server): Add response type detecting to lang redirect

Commit:
Feature(settings): complete settings page logic

Commit:
Fix(settings): typo

Commit:
Feature(settings): update language settings copy

Commit:
Feature(settings): Update url/challenge lang onChange

Commit:
Feature(challenges): Update title on challenge change

Commit:
Feature(analytics): Add redux logic for analytics
Add(nav): Add event tracking to nav bar
Add(Drawer): Add event tracking to chat/map drawer

raisedadead committed:
Fix (settings) : Duplicate social labels

closes #9848

Commit:
Fix: remove old middleware config

Commit:
Feature(challenges): Load and cache required files

Commit:
chore(package): Update to react-bootstrap@0.30
This required a little of refactoring to get working
correctly.

Commit:
Feature(challenges): add required to challenge creationed

Commit:
Feature(challenges): challenges define their own requires

Commit:
Fix(routes): ignore link routes in lang redirects

Commit:
Fix(map): map drawer new tab button

Commit:
Fix(settings): remove underline in update-email button

Commit:
Fix(css): trimming of app sides on mobile
add class container to body

Commit:
Fix(layout): use specific body css tag for react

Commit:
Fix(scripts): make sure rev-manifest exists while running lint

Commit:
Fix(scripts): run create-rev on pretest as well

raisedadead commented:
@BerkeleyTrue
###  :sparkles: AWESOME! YOU SIR ARE THE ROCKSTAR :sparkles:

atjonathan commented:
@BerkeleyTrue :heart:
